dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/option/OptionGroupDeletionHandler.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2004-2022, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.option;
+
+import java.util.List;
+
+import lombok.AllArgsConstructor;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.hisp.dhis.system.deletion.DeletionHandler;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+@AllArgsConstructor
+public class OptionGroupDeletionHandler extends DeletionHandler
+{
+    @Autowired
+    private OptionGroupStore optionGroupStore;
+
+    @Override
+    protected void register()
+    {
+        whenDeleting( Option.class, this::deleteOption );
+    }
+
+    private void deleteOption( Option option )
+    {
+        List<OptionGroup> optionGroup = optionGroupStore.getOptionGroups( option.getUid() );
```

Comment:
I think ideally we could use a query which fetches only the option groups which contains the given option, as opposed to retrieving all option groups and iterating in memory. 

Typically there will not be a high number of option groups, but as we know deletion can be slow, so as a good principle we should try to pull only the necessary objects from the database vs all.

Reply:
The method `optionGroupStore.getOptionGroups( option.getUid() )` queries only `optionGroup` which contains given option ID. I changed the method name to `getOptionGroupsByOptionId()` to make it less confusing.